packages/dds/tree/docs/repair-data/README.md

```diff
@@ -0,0 +1,268 @@
+It should be updated once more progress is made on the implementation.
+
+## What is Repair Data
+
+In order to undo an edit `e` that was applied to state `s1` and yielded state `s2`,
+we need to construct an inverse edit `e⁻¹` that yields state `s1` when applied to state `s2`.
+Computing `e⁻¹` requires the following information:
+
+-   The original change to be undone
+-   Any document state that we wish to restore as part of the undo
+    but cannot derive from the the original change.
+
+This last bullet point may not be immediately obvious,
+but it is a critical issue.
+We may need to recover information about the state of the document before the change that we wish to undo
+because document changes can be destructive:
+
+-   Setting the value on a node erases information about the prior value of that node.
+-   Deleting a subtree erases information about the contents of that subtree.
+
+Such data cannot be derived solely from the original changeset to be undone.
+We refer to the information that is needed in addition the original changeset as "repair data".
+
+## Three Contexts
+
+Repair data is used in three different contexts.
+
+### Rolling Back Local Transactions
+
+When a client needs to edit the document, it runs a command as part of a transaction.
+Commands are allowed to return a special value to communicate to the transaction code that the transaction should be aborted.
+When that happens, any edits that the command had already applied to the document state need to be rolled back.
+
+Rolling back changes could be achieved by editing a separate copy of the document during the transaction
+(possibly though a persistent data structure, or a copy on write system).
+The current implementation (which will likely change soon) edits the one true document state as the transaction progresses, and rolls back the changes afterward.
+It actually rolls back the changes even if the transaction succeeds,
+but that's not particularly relevant to our discussion.
+
+### Local Branch Updates
+
+When a client successfully completes a local transaction,
+it updates the local document state to reflect the impact of the edit,
+and it sends the edit to the service for sequencing.
+Under ideal circumstances, the next edit that the client receives from the service is that same edit that it had applied locally and sent for sequencing.
+If that's the case, the client does not need to update the document state.
+It's possible for an edit from a peer to be sequenced before the edit that was applied locally and sent out.
+When that happens,
+the local client needs to update the document state to reflect not only the impact of the edit from the peer,
+but also the impact that the peer edit has on the local edit.
+
+For example,
+if a the local edit was deletion of two nodes A and B,
+with a constraint that both nodes must exist,
+and the peer change had deleted node A,
+then the net impact of the peer change is that node B ought to be revived.
```

Comment:
I'm having trouble understanding this example. If the constraint is that A and B must both exist, does this not mean that the local change should only be applied if they both exist? So, if the peer change deletes A prior to the local change, then the local change should not be applied.

While typing this out, I realized that node B needs to be revived so that the state reflects the local change not being applied. There was some confusion though, so I wonder if there's a better way to word this.

Reply:
Let me know if it's clear now.

Reply:
Yup, it's clear!